web3/providers/async_base.py

```diff
@@ -40,53 +48,61 @@
     )
 
 
+CACHEABLE_REQUESTS = cast(
+    Set[RPCEndpoint],
+    (
+        "eth_chainId",
+        "eth_getBlockByHash",
+        "eth_getBlockTransactionCountByHash",
+        "eth_getRawTransactionByHash",
+        "eth_getTransactionByBlockHashAndIndex",
+        "eth_getTransactionByHash",
+        "eth_getUncleByBlockHashAndIndex",
+        "eth_getUncleCountByBlockHash",
+        "net_version",
+        "web3_clientVersion",
+    ),
+)
+
+
 class AsyncBaseProvider:
-    _middlewares: Tuple[AsyncMiddleware, ...] = ()
-    # a tuple of (all_middlewares, request_func)
     _request_func_cache: Tuple[
-        Tuple[AsyncMiddleware, ...], Callable[..., Coroutine[Any, Any, RPCResponse]]
-    ] = (
-        None,
-        None,
-    )
+        Tuple[Middleware, ...], Callable[..., Coroutine[Any, Any, RPCResponse]]
+    ] = (None, None)
 
     is_async = True
     has_persistent_connection = False
     global_ccip_read_enabled: bool = True
     ccip_read_max_redirects: int = 4
 
-    @property
-    def middlewares(self) -> Tuple[AsyncMiddleware, ...]:
-        return self._middlewares
+    # request caching
+    cache_allowed_requests: bool = False
+    cacheable_requests: Set[RPCEndpoint] = CACHEABLE_REQUESTS
+    _request_cache: SimpleCache
+    _request_cache_lock: asyncio.Lock = asyncio.Lock()
 
-    @middlewares.setter
-    def middlewares(self, values: MiddlewareOnion) -> None:
-        # tuple(values) converts to MiddlewareOnion -> Tuple[Middleware, ...]
-        self._middlewares = tuple(values)  # type: ignore
+    def __init__(self) -> None:
+        self._request_cache = SimpleCache(1000)
 
     async def request_func(
-        self, async_w3: "AsyncWeb3", outer_middlewares: AsyncMiddlewareOnion
+        self, async_w3: "AsyncWeb3", middlewares: MiddlewareOnion
     ) -> Callable[..., Coroutine[Any, Any, RPCResponse]]:
         # type ignored b/c tuple(MiddlewareOnion) converts to tuple of middlewares
-        all_middlewares: Tuple[AsyncMiddleware] = tuple(outer_middlewares) + tuple(self.middlewares)  # type: ignore  # noqa: E501
+        middlewares: Tuple[Middleware, ...] = tuple(middlewares)  # type: ignore
```

Comment:
nit: does this `type: ignore` still need to be here?

Reply:
Unfortunately yeah. I played around with this a bit but didn't get too far. It comes fromt he fact that the `__iter__()` method defined before my time will turn the onion into a `Tuple[Middleware, ...]` but to type it as such isn't straightforward. The type here _is_ correct, we just don't have a very elegant solution for this (yet) and maybe we can tackle that as a separate issue altogether. Could be worth looking into. I think it has to do with satisfying the `Mapping` inheritance for the `NamedElementOnion`.

Reply:
Just kidding. The moment I said a definitive yeah something clicked. It was easily solved by creating a method to tupelize the onion into a tuple of middlewares. I pushed that up as a separate commit in case that isn't ideal but I prefer than to the `type: ignore`s myself. Let me know what you think on that.

Reply:
Nice! Yeah, I think I prefer that too